### DISCUSSION TITLE
How to get requirements with extras?

### DISCUSSION THREAD

**Discussion by MrMarvel** [pypa/setuptools#4052] [Q&A] [answered]
```python
from pkg_resources import working_set, get_distribution, VersionConflict, DistributionNotFound, DistInfoDistribution

for dist in working_set:
    dist: DistInfoDistribution
    dist.requires()
```
I can get requirements with method requires.
But I want to **_get all requirements with extras_**.
I can get extras names as str with property extras but not their requirements
EDIT: I need to get information from extra package requirements.
```python
for dist in working_set:
    dist: DistInfoDistribution
    extras: list[str] = dist.extras
```

**Comment by abravalheri** [comment]
If you use `packaging.requirements.Requirement` to parse each item `dist.requires` you should be able to access [`req.extras`](https://packaging.pypa.io/en/stable/requirements.html) (from there you can use things like `filter`, `itertools.groupby` or similar...).

Also please note that `pkg_resources` is deprecated, users are advised to migrate to [`importlib.metadata/importlib_metadata`](https://importlib-metadata.readthedocs.io/en/latest/using.html).

**Reply by MrMarvel** [reply]
I want to find requirements without any additionals packages installed (minimal python (pip setuptools))
packaging
`from packaging.requirements import Requirement` requires additional package

**Reply by MrMarvel** [reply]
Also I don't understand why does packaging.requirements.Requirement help.
packaging only help to to format string requirements to class object with api access with no parsing requirements so there is no new information I can get from it. req.extras just returns str as I put in Requirement declaration.

![image](https://github.com/pypa/setuptools/assets/24212110/dd679096-dcd9-435a-9510-50f8248d46a3)

Still idk how to get extra requirements

**Comment by MrMarvel** [comment]
I found out something interesting.
Strange extras list generator
![image](https://github.com/pypa/setuptools/assets/24212110/6a484a45-63de-4103-a70e-6689f54769ed)

**Comment by MrMarvel** [comment, marked as answer]
Whoila!
`working_set.by_key['jsonschema'].requires(('format',))` returns `[Requirement.parse('attrs>=22.2.0'), Requirement.parse('jsonschema-specifications>=2023.03.6'), Requirement.parse('referencing>=0.28.4'), Requirement.parse('rpds-py>=0.7.1'), Requirement.parse('fqdn; extra == "format"'), Requirement.parse('idna; extra == "format"'), Requirement.parse('isoduration; extra == "format"'), Requirement.parse('jsonpointer>1.13; extra == "format"'), Requirement.parse('rfc3339-validator; extra == "format"'), Requirement.parse('rfc3987; extra == "format"'), Requirement.parse('uri-template; extra == "format"'), Requirement.parse('webcolors>=1.11; extra == "format"')]`
